File: ROID-Forum-Server/ROID-Forum-Server/Convert Server Code/GMGServer/bin/Server/Chat Box/ChatController.dart
library ChatController;
import '../Server.dart';
import '../User.dart';
import 'dart:convert';
import '../User.dart';
import 'ChatIOController.dart';

class ChatController {
  List<String> chats = new List();
  Server server;
  List<User> users = new List();
  ChatIOController chatIOController;
  ChatController(Server s) {
    chatIOController = new ChatIOController(this);
    server = s;
    chatIOController.loadAllChats();
  }
  void addUser(User u) {
    users.add(u);
    sendListUpdateToAll();
    sendAllChats(u);
  }
  void removeUser(User u) {
    users.remove(u);
    sendListUpdateToAll();
  }
  void sendListUpdateToAll() {
    String list = "Online(${users.length}): ";
    int guests = 0;
    bool addComma = false;
    for (User u in users) {
      if (u.account != null) {
        if (addComma) {
          list += ", ";
        } else {
          addComma = true;
        }
        list += u.account.name;
      } else {
        guests += 1;
      }
    }
    if (addComma) {
      list += " and ${guests} Guests";
    } else {
      list += "${guests} Guests";
    }

    Map m = new Map();
    m['Controller'] = 'Chat';
    m['Title'] = 'Online List';
    m['Data'] = list;
    for (User u in users) {
      u.sendMap(m);
    }
    print(list);
  }
  void logic() {
    chatIOController.logic();
  }
  void addChat(User u, String chat) {
    chat = '${u.account.name}: ' + chat;
    chats.add(chat);
    //Send chat to all connected websockets
    Map m = new Map();
    m['Controller'] = 'Chat';
    m['Title'] = 'Msg';
    m['Data'] = chat;
    String message = JSON.encode(m);
    for (int i = 0; i < server.networkingController.users.length; i++) {
      User u = server.networkingController.users[i];
      u.sendString(message);
    }
  }
  void sendAllChats(User u) { //Send only last 20
    int start = 0;
    // if (chats.length > 20) {
    //   start = chats.length - 20;
    // }
    for (int i = start; i < chats.length; i++) {
      Map m = new Map();
      m['Controller'] = 'Chat';
      m['Title'] = 'Msg';
      m['Data'] = chats[i];
      u.sendMap(m);
    }
  }
  void onMessage(User u, Map message) {
    if (message['Title'] == 'Msg' && u.account != null) {
      addChat(u, message['Data']);
    }
  }
}

File: ROID-Forum-Server/ROID-Forum-Server/Chat Box/ChatController.cs
using System;
using System.Collections.Generic;
using Cassandra;

namespace ROIDForumServer
{
    public class ChatController(ServerController serverController)
    {
        // Todo: We can use cassandra, redis, or nats to properly show
        // active users and allow subscribing them to events.
        // This will allow multi-instance scaling of the server.
        private readonly List<ConnectedUser> _users = new List<ConnectedUser>();
        private ISession DatabaseSession { get; } = serverController.Database.GetSession();
        private Networking Networking { get; } = serverController.Networking;
        public void AddUser(ConnectedUser user)
        {
            _users.Add(user);
            SendListUpdateToAll();
            SendAllChats(user);
        }
        public void RemoveUser(ConnectedUser user)
        {
            _users.Remove(user);
            SendListUpdateToAll();
        }
        public void SendListUpdateToAll()
        {
            String list = $"Online({_users.Count}): ";
            int guests = 0;
            bool addComma = false;
            // Todo: Fix bulk account name lookups, either cache names
            // or make an active online table in Cassandra, or
            // make a bunch of async parallel queries to be distributed
            // among cassandra nodes.
    foreach (ConnectedUser user in _users)
            {
                if (user.AccountId != null)
                {
                    if (addComma)
                    {
                        list += ", ";
                    }
                    else
                    {
                        addComma = true;
                    }
                    list += DatabaseAccount.GetAccountName(DatabaseSession, (Guid)user.AccountId);
                }
                else
                {
                    guests += 1;
                }
            }
            if (addComma)
            {
                list += $" and {guests} Guests";
            }
            else
            {
                list += $"{guests} Guests";
            }

            byte[] message = ServerMessages.ChatOnlineList(list);
            foreach (ConnectedUser user in _users)
            {
                user.Send(message);
            }
            Console.WriteLine(list);
        }
        private void AddChat(ConnectedUser user, String chat)
        {
            if (user.AccountId == null)
            {
                return;
            }
            chat = $"{DatabaseAccount.GetAccountName(DatabaseSession, (Guid)user.AccountId)}: " + chat;
            DatabaseChat.SubmitChat(DatabaseSession, (Guid)user.AccountId, chat);
            //Send chat to all connected websockets
            // Todo: Switch to a NATS subscription model
            byte[] chatMsg = ServerMessages.ChatMessage(chat);
            foreach (var user2 in Networking.Users)
            {
                user2.Send(chatMsg);
            }
        }
        private void SendAllChats(ConnectedUser user)
        {
            foreach (var (_, _, content, _) in DatabaseChat.GetRecentChats(DatabaseSession))
            {
                user.Send(ServerMessages.ChatMessage(content));
            }
        }
        public void OnMessage(ConnectedUser user, Dictionary<string, object> message)
        {
            if ((string)message["Title"] == "Msg" && user.AccountId != null)
            {
                AddChat(user, (string)message["Data"]);
            }
        }
    }
}